Remove reject ACLs regardless of cache

trozet opened the pull request:
Previously we were just checking if there as a known ACL which existed
in the cache. If the cache is ever out of sync, we could accidentally
leave stale reject ACLs configured in OVN.

This patch adds checking during service sync for any known stale ACLs
with reject actions belonging to a service. During regular reject ACL
delete a new check is added to see if cache is in an invalid state, and
if so attempt to remove the ACL by querying OVN.

Signed-off-by: Tim Rozet <trozet@redhat.com>

trozet commented:
/hold

Want to test this more and maybe also add another commit for handling stale ACL removal in syncServices.

alexanderConstantinescu commented on go-controller/pkg/util/net.go:
```diff
@@ -139,6 +139,23 @@ func JoinHostPortInt32(host string, port int32) string {
 	return net.JoinHostPort(host, strconv.Itoa(int(port)))
 }
 
+// SplitHostPortInt32 splits a vip into its host and port counterparts
```
You could use: https://golang.org/pkg/net/#SplitHostPort ?

trozet replied:
done

aojea commented on go-controller/pkg/ovn/endpoints.go:
```diff
@@ -43,7 +43,7 @@ func (ovn *Controller) getLbEndpoints(ep *kapi.Endpoints) map[kapi.Protocol]map[
 
 // AddEndpoints adds endpoints and creates corresponding resources in OVN
 func (ovn *Controller) AddEndpoints(ep *kapi.Endpoints) error {
-	klog.V(5).Infof("Adding endpoints: %s for namespace: %s", ep.Name, ep.Namespace)
```
I think that the logging should be changed separately and make it consistent in all components, actually, it will be better if we define some logging policies and enforce them in all the code.
Maybe define a default loglevel (i.e. `klog.V(2).Infof`) and set it by default , so users can decide if they want to swith it off later?

trozet replied:
Right now we only use v(5) for debug, and default Infof logging. These logging messages should not have been at debug. We log this level of info at default log level for other things like pods. If we want to use other log levels or standardize things, we can do it in another PR.

aojea replied:
Fair enough

aojea commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -267,12 +267,13 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb string, sourceIP string, s
 }
 
 func (ovn *Controller) deleteLoadBalancerRejectACL(lb, vip string) {
-	acl, _ := ovn.getServiceLBInfo(lb, vip)
```
with this change seems that ` ovn.getServiceLBInfo(lb, vip)` does no longer needs to return the acl.
Is it always better to generate it from the lb,vip?

trozet replied:
I modified this to still try to use the cache, and if it finds the cache is invalid then last resort is try to query OVN. So we still need this. See upcoming change.

dcbw commented:
If we have to do some kind of "sync OVN database to Kube API state" when ovnkube restarts, we should do that in the sync functions if possible.

trozet commented:
> If we have to do some kind of "sync OVN database to Kube API state" when ovnkube restarts, we should do that in the sync functions if possible.

@dcbw I added some handling in the services sync function. This is a little tricky because these are ACLs, so we can't be 100% sure that they belong to services unless the name matches. Therefore I only blow those away if I can tell that they have the expected name of a service and are stale (aka have endpoints). I fixed the regular delete operation to use the cache. However, when we can tell the cache is in an invalid state, last resort is still to try to query OVN.

trozet commented:
running downstream on https://github.com/openshift/ovn-kubernetes/pull/295

dcbw commented on go-controller/pkg/ovn/service.go:
```diff
@@ -48,6 +52,16 @@ func (ovn *Controller) syncServices(services []interface{}) {
 
+		// detect if service has endpoints for stale reject ACL check. If there are endpoints, we need to wipe any
+		// old stale ACLs
+		ep, err := ovn.watchFactory.GetEndpoint(service.Namespace, service.Name)
```
Does this go directly to the apiserver, or to the informer cache? If it's the apiserver directly, then it'll potentially take a long time to read all of them on startup. If it's the cache, since WatchServices() gets called before WatchEndpoints(), this will always return nothing.

dcbw replied:
After discussion with Tim I'm probably wrong, because we call WaitForCacheSync() which /should/ populate all the caches with the initial set before running any of the Sync handlers.

dcbw commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -281,17 +312,25 @@ func (ovn *Controller) deleteLoadBalancerRejectACL(lb, vip string) {
+		}
+	}
+
+	if len(args) > 0 {
```
@trozet maybe I'm misreading the diff, but isn't this block redundant with the one just above at 318?

trozet replied:
oops copy pasta error. I'll remove that. thank.

dcbw commented on go-controller/pkg/ovn/service.go:
```diff
@@ -57,14 +71,101 @@ func (ovn *Controller) syncServices(services []interface{}) {
+				if len(entry) != 2 {
+					continue
+				}
+				name := entry[0].(string)
```
@trozet I'm a bit worried about cast assertions here if things go wrong with the JSON. Can we have some "ok" stuff here and double-check that our cases are OK so we don't panic?

trozet replied:
yeah, I didn't even know you can do that in golang. Will add

dcbw commented:
/lgtm
but the testcase failures both seem service-related; do you think they are relevant?

girishmg commented on go-controller/pkg/ovn/service.go:
```diff
@@ -57,14 +71,107 @@ func (ovn *Controller) syncServices(services []interface{}) {
+									gatewayRouter, err)
+								continue
+							}
+							name := ovn.generateACLName(lb, physicalIP, svcPort.NodePort)
```
L85: Not sure why you are checking for `err` inside the for loop here. I initially thought that maybe you wanted to print a warning for every physical IP, but the error message is not printing the physicalIP. So, should you move this check outside of the for loop?

trozet replied:
copy paste error from when I was copying this logic from createService. Thanks for catching it. I'll post a fix shortly.

trozet commented:
> /lgtm
> but the testcase failures both seem service-related; do you think they are relevant?

I didn't look closely. I saw there were a few flakes, and that's about what we usually get out of our CI these days :1st_place_medal: 

Let's see what the next round of CI results bring.

trozet commented:
I think CI looks OK. These changes aren't IPv6 or shared gw specific.
